resourcetopo/manager.go

```diff
@@ -0,0 +1,166 @@
+
+	relationEventQueue chan relationEvent
+	nodeEventQueue     chan nodeEvent
+}
+
+func NewResourcesTopoManager(cfg Config) (Manager, error) {
+	if err := checkConfig(&cfg); err != nil {
+		return nil, err
+	}
+	m := &manager{
+		storages:           make(map[string]*nodeStorage),
+		nodeEventQueue:     make(chan nodeEvent, cfg.NodeEventQueueSize),
+		relationEventQueue: make(chan relationEvent, cfg.RelationEventQueueSize),
+	}
+
+	if err := m.AddConfig(cfg); err != nil {
+		return nil, err
+	}
+
+	return m, nil
+}
+
+func (m *manager) AddConfig(cfg Config) error {
+	for _, r := range cfg.Resolver {
+		preOrderStorage, err := m.getOrCreateStorage(r.PreOrder, cfg.GetInformer)
+		if err != nil {
+			return err
+		}
+		for _, v := range r.PostOrders {
+			_, err := m.getOrCreateStorage(v, cfg.GetInformer)
+			if err != nil {
+				return err
+			}
+		}
+		if err = preOrderStorage.addRelationConfig(&r); err != nil {
+			return err
+		}
+	}
+
+	for _, d := range cfg.Discoverer {
+		s, err := m.getOrCreateStorage(d.PostMeta, cfg.GetInformer)
+		if err != nil {
+			return err
+		}
+		_, err = m.createVirtualStorage(d.PreMeta)
+		if err != nil {
+			return err
+		}
+
+		if err = s.addDiscoverConfig(&d); err != nil {
+			return err
+		}
+	}
+
+	return m.dagCheck()
+}
+func (m *manager) AddObjectHandler(meta metav1.TypeMeta, handler NodeHandler) error {
+	s := m.storages[generateMetaKey(&meta)]
+	if s == nil {
+		return fmt.Errorf("resource %v not configured in this manager", meta)
+	}
+	return s.AddObjectHandler(handler)
+}
+
+func (m *manager) GetTopo(meta metav1.TypeMeta, name types.NamespacedName) (NodeInfo, error) {
+	s := m.storages[generateMetaKey(&meta)]
+	if s == nil {
+		return nil, fmt.Errorf("resource %v not configured in this manager", meta)
+	}
+	return s.GetNode(&name), nil
+}
+
+func (m *manager) GetResourceTopo(object metav1.TypeMeta) TopoNodeStorage {
+	return m.storages[generateMetaKey(&object)]
+}
+
+func (m *manager) getOrCreateStorage(typeMeta metav1.TypeMeta, getInformer func(meta metav1.TypeMeta) cache.SharedInformer) (*nodeStorage, error) {
+	key := generateMetaKey(&typeMeta)
+
+	s, ok := m.storages[key]
+	if ok {
+		return s, nil
+	}
+
+	informer := getInformer(typeMeta)
+	if informer == nil {
+		return nil, fmt.Errorf("failed to get informer for resource %s", key)
+	}
+
+	s = newNodeStorage(m, informer, typeMeta)
+	m.storages[key] = s
+	return s, nil
+}
+
+func (m *manager) getStorage(meta metav1.TypeMeta) *nodeStorage {
+	key := generateMetaKey(&meta)
+
+	return m.storages[key]
+}
+
+func (m *manager) createVirtualStorage(object metav1.TypeMeta) (*nodeStorage, error) {
+	key := generateMetaKey(&object)
+	_, ok := m.storages[key]
+	if ok {
+		return nil, fmt.Errorf("unexpected twice configuration for virtual resource %s", key)
+	}
+
+	s := newVirtualStorage(m, object)
+	m.storages[key] = s
+	return s, nil
+}
+
+func (m *manager) AddRelationHandler(preOrder, postOrder metav1.TypeMeta, handler RelationHandler) error {
```

Comment:
We can add some comments for those key methods.

Reply:
added in types interface define